Added is_protected_parameter function and check for post taxonomy query building

brandon-m-skinner opened the pull request:
[PLAT-977](https://vipjira.atlassian.net/browse/PLAT-977)

~~Needed to allow the filtering of taxonomies due to the discovery that
certain taxonomies, like post_status, cause weird interactions with
ElasticPress and cause queries against Elasticsearch to always miss and
return nothing.~~

Added is_protected_parameter function and check for post taxonomy query building. This fixes the issue we were seeing with post_status in our health checks and should prevent this issue in the future so long as the protected parameters are kept up to date.

With this and only this change, the health check is fixed. No mu-plugins changes required 🎉.

Commit:
Added ep_post_get_taxonomies filter

Needed to allow the filtering of taxonomies due to the discovery that
certain taxonomies, like post_status, cause weird interactions with
ElasticPress and cause queries against Elasticsearch to always miss and
return nothing

Commit:
Add parentheses to get_taxonomies call

nickdaugherty commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -771,7 +771,15 @@ public function format_args( $args, $wp_query ) {
+		 * @param  {array} $taxonomies an array of taxonomies
+		 * @return  {array} List of taxonomies to use for tax query
+		 */
+		$taxonomies = apply_filters( 'ep_post_get_taxonomies', get_taxonomies() );
```
The Post indexable has a method that we can probably reuse here - `get_indexable_post_taxonomies()` - it's used when generating an ES doc for indexing...so presumably it would make sense to limit which taxonomies we try to parse out of the args to the same ones?

Does introduce a bit of a disconnect between the `WP_Query()` and the _actual_ ES query that gets created, but I think if it's limited to being used in only this part that tries to parse out taxonomies from the root of the WP_Query, it should be safe.

Aside - `get_indexable_post_taxonomies()` also has a similar filter - `ep_sync_taxonomies`.

brandon-m-skinner replied:
@nickdaugherty 

Sure, makes sense. I was doing the change in a way to get it merged upstream with minimal changes(they just use get_taxonomies). 

I'll update here and then apply it to mu-plugins after :+1:

nickdaugherty replied:
Actually not so sure that's the best way, because ideally we can still allow those taxonomies to be indexed (not filter them out). That would be particularly important in the case of Edit Flow (where the `post_status` taxonomy comes from).

Instead, we can modify this behavior to not interpret any "core" query arguments as taxonomies...so basically have a list of the known WP_Query vars (like `post_status`) and then EP doesn't try to interpret them as taxonomies, even if there is a taxonomy with that name.

nickdaugherty replied:
Full list of core params: https://developer.wordpress.org/reference/classes/wp_query/#parameters

nickdaugherty reviewed (changes requested):
Nice find - I noticed that the Post indexable already has a method/filter for limiting taxonomies, it probably makes sense to re-use that one here.

Commit:
Switch from filter to comparison

Since filtering would affecting more than just whether something is
treated as a taxonomy or not, a new approach was needed. It made sense
to just build a list of protected parameters based on WP_Query and then
compare individual taxonomies to it.

Commit:
Merge conflict

brandon-m-skinner commented:
Closed the mu-plugins PR and deleted the branch since it's no longer needed. This change fixes our issue entirely by itself.

Updated PR content and such to keep history rather than make a new one.

nickdaugherty commented on includes/classes/Indexable/Post/Post.php:
```diff
@@ -1625,4 +1625,88 @@ protected function get_orderby_array( $orderbys ) {
+			'year',
+		);
+
+		return in_array( $value, $protected_parameter_list );
```
```suggestion
		return in_array( $value, $protected_parameter_list, true );
```

Minor - it's usually best to use the "strict" mode of in_array() by passing `true` as the 3rd param

brandon-m-skinner replied:
With the string type check, only strings will make it that far right now.

However, future proofing is a good idea and you already typed it out :+1:.

nickdaugherty reviewed (approved):
Looks good, just left one minor inline suggestion.

Commit:
Update includes/classes/Indexable/Post/Post.php

Co-Authored-By: Nick Daugherty <ndaugherty987@gmail.com>

brandon-m-skinner commented:
r826-stacks

nickdaugherty commented:
More context for future reference: https://a8c.slack.com/archives/C013N7TDZAT/p1638203936171200?thread_ts=1637152968.437500&cid=C013N7TDZAT
